### PR TITLE
Share image

### LINKED ISSUE #10: Share Image
#### As a user I want to share an image with someone via email.

Acceptance criteria:
- [ ] There is a link to share an image.
- [ ] After clicking on this link, a bootstrap modal appears with a form to
  send the email.
- [ ] The form has fields for the recipient's email and an optional message.
- [ ] The email we send has an HTML part and a text part.
- [ ] The email contains the image, the custom message if provided, and a link
  inviting the recipient to check out our application.
- [ ] The modal closes and a success message is shown if the email was sent
  successfully.
- [ ] The user is alerted of errors if the operation was unsuccessful.

Discussion topics:
- [ ] How could we develop the front end and back end separately? (And why
  might that be a good idea?)
- [ ] How should we implement a non-CRUD action?
- [ ] What are the different types of errors we should handle?
- [ ] What is AJAX? What libraries can help us add AJAX to our Rails app?
- [ ] How should we go about sending emails from our Rails app?
- [ ] How could we show a preview of the image being shared in the modal?
- [ ] What are the implications of allowing users to send emails from our app?

Reference material:
- http://v4-alpha.getbootstrap.com/components/modal
- http://guides.rubyonrails.org/action_mailer_basics.html
- https://elements.heroku.com/addons/sendgrid
- https://devcenter.heroku.com/articles/sendgrid#ruby-rails
- https://github.com/rails/jquery-ujs/wiki

Dependencies:
- Image Index

### PR DESCRIPTION
**Pull request by tlconnor**
Closes #10

### CODE REVIEW

**Comment by bboe** [comment]
Reassigned this issue to @rmacklin

File: config/environment.rb [hunk 1]
```diff
@@ -1,5 +1,15 @@
...
 
+ActionMailer::Base.smtp_settings = {
+  address:        'smtp.sendgrid.net',
+  port:           '587',
+  authentication: :plain,
+  user_name:      ENV['SENDGRID_USERNAME'],
+  password:       ENV['SENDGRID_PASSWORD'],
+  domain:         'heroku.com',
+  enable_starttls_auto: true
+}
```
Comment (rmacklin) [inline config/environment.rb]: Does this belong in the production environment?

File: config/environments/development.rb [hunk 2]
```diff
@@ -30,6 +30,7 @@
   config.action_mailer.raise_delivery_errors = false
 
   config.action_mailer.perform_caching = false
+  config.action_mailer.default_url_options = { host: 'tlconnor.herokuapp.com' }
```
Comment (rmacklin) [inline config/environments/development.rb]: It would be better to depend on an environment variable for this, so you don't need to redeploy if you change your subdomain

File: config/environments/production.rb [hunk 3]
```diff
@@ -50,7 +50,9 @@
 
   # Ignore bad email addresses and do not raise email delivery errors.
   # Set this to true and configure the email server for immediate delivery to raise delivery errors.
-  # config.action_mailer.raise_delivery_errors = false
+  config.action_mailer.raise_delivery_errors = true
+
+  config.action_mailer.default_url_options = { host: 'tlconnor.herokuapp.com' }
```
Comment (rmacklin) [inline config/environments/production.rb]: This change doesn't seem correct. We'd want local emails link to localhost

File: test/mailers/images_mailer_test.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+  test 'share_image includes a text/html part' do
+    email = ImagesMailer.share_image(@image, 'me@example.com', 'example message')
+    email.deliver_now
+
+    part = email.parts[1]
+    assert_match %r{text/html}, part.content_type
+    assert_match('<p>example message</p>', part.to_s)
+    assert_match(%(<img src="#{@image.url}">), part.to_s)
+  end
```
Comment (rmacklin) [inline test/mailers/images_mailer_test.rb]: You should test the conditional logic in the email

File: app/views/images_mailer/share_image.html.erb [hunk 5]
```diff
@@ -0,0 +1,11 @@
+<h1>An image has been shared with you</h1>
```
Comment (rmacklin) [inline app/views/images_mailer/share_image.html.erb]: If you want to guarantee this heading stays in sync with the email subject, you should use another instance variable.

File: app/validators/image_share_validator.rb [hunk 6]
```diff
@@ -0,0 +1,9 @@
+class ImageShareValidator
+  include ActiveModel::Model
+  include Virtus.model
+
+  attribute :recipient, String
+  attribute :message, String
+
+  validates :recipient, presence: true, format: { with: /.+@.+\..+/i, allow_blank: true }
+end
```
Comment (rmacklin) [inline app/validators/image_share_validator.rb]: Do we actually need Virtus? It seems like all of this could be accomplished with just ActiveModel.

File: app/validators/image_share_validator.rb [hunk 7]
```diff
@@ -0,0 +1,9 @@
+class ImageShareValidator
```
Comment (rmacklin) [inline app/validators/image_share_validator.rb]: This is really a form object (containing a validation), not a validator. So `ImageShareForm` would be a more appropriate name.

File: test/validators/image_share_validator_test.rb [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+    validator = ImageShareValidator.new(recipient: 'invalid recipient')
+    assert_predicate validator, :invalid?
+    assert_equal ['is invalid'], validator.errors['recipient']
+  end
+
+  def test_valid_recipient
+    validator = ImageShareValidator.new(recipient: 'nobody@appfolio.com')
+    assert_predicate validator, :valid?
+  end
+end
```
Comment (rmacklin) [inline test/validators/image_share_validator_test.rb]: To channel @bboe, this should test some more valid and invalid cases to cover the whole regex.

File: app/controllers/images_controller.rb [hunk 9]
```diff
@@ -1,15 +1,29 @@
...
+    validator = ImageShareValidator.new(params.permit(:recipient, :message))
+    validator.validate!
+    ImagesMailer.share_image(@image, validator.recipient, validator.message).deliver_now
+    head :no_content
+  end
+
+  protected
+
+  def fetch_image
+    @image = Image.find(params[:id]) if params[:id]
```
Comment (rmacklin) [inline app/controllers/images_controller.rb]: I think it's better to explicitly declare the actions you want this `before_action` to apply to than to have it be implicit based on `if params[:id]`

File: app/controllers/images_controller.rb [hunk 10]
```diff
@@ -1,15 +1,29 @@
...
   end
+
+  def share
+    validator = ImageShareValidator.new(params.permit(:recipient, :message))
+    validator.validate!
+    ImagesMailer.share_image(@image, validator.recipient, validator.message).deliver_now
+    head :no_content
+  end
+
+  protected
```
Comment (rmacklin) [inline app/controllers/images_controller.rb]: There's no reason to use `protected` here. It should be `private`. If you're not familiar, `protected` in Ruby is different than in Java/other languages; it's _very rare_ you'll need to use `protected` in Ruby.

File: test/controllers/images_controller_test.rb [hunk 11]
```diff
@@ -68,4 +68,33 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
...
+        message: 'here is the image for you'
+      }
+    end
+
+    assert_equal 204, response.status
+
+    email = ActionMailer::Base.deliveries.last
+    assert_equal ['ropes@appfolio.com'], email.from
+    assert_equal ['nobody@appfolio.com'], email.to
+    assert_equal 'An image has been shared with you', email.subject
```
Comment (rmacklin) [inline test/controllers/images_controller_test.rb]: We should also assert that the _correct_ image was shared.

Comment (rmacklin) [inline test/controllers/images_controller_test.rb]: And that the message we sent was included in the email

File: test/controllers/images_controller_test.rb [hunk 12]
```diff
@@ -68,4 +68,33 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
     assert_equal 422, response.status
     assert_equal ['is not a valid image URL'], body['errors']['url']
   end
+
+  test 'share should send an email' do
+    image = Image.create! url: 'http://example.image.url'
+
+    assert_difference -> { ActionMailer::Base.deliveries.length }, 1 do
+      post share_image_url(image), as: :json, params: {
```
Comment (rmacklin) [inline test/controllers/images_controller_test.rb]: Your app sends ajax requests right? All of these controller tests should be using `xhr`

File: app/controllers/images_controller.rb [hunk 13]
```diff
@@ -1,15 +1,29 @@
...
   def create
     @image = Image.new(params.permit(:url))
     @image.save!
     render status: 201
   end
+
+  def share
+    validator = ImageShareValidator.new(params.permit(:recipient, :message))
+    validator.validate!
+    ImagesMailer.share_image(@image, validator.recipient, validator.message).deliver_now
```
Comment (rmacklin) [inline app/controllers/images_controller.rb]: What will happen if the image wasn't found?

File: test/controllers/images_controller_test.rb [hunk 14]
```diff
@@ -68,4 +68,33 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
...
+  test 'share should validate the passed recipient' do
+    image = Image.create! url: 'http://example.image.url'
+
+    post share_image_url(image), as: :json, params: { recipient: '' }
+
+    body = JSON.parse(response.body)
+    assert_equal 422, body['status']
+    assert_equal 422, response.status
+    assert_equal ["can't be blank"], body['errors']['recipient']
+  end
```
Comment (rmacklin) [inline test/controllers/images_controller_test.rb]: Related to https://github.com/AppFolioOnboarding/image-sharer-tlconnor/pull/29#discussion_r74321133, you should always test edge cases. Specifically, any member actions could be hit with an invalid id, and we want to test that case.

**Comment by rmacklin** [comment]
A more descriptive commit message for fb2d90e615f3470584a95cfcee94196aa33b31c0 would be "add share action and route"

File: app/controllers/images_controller.rb (hunk 13 above)
Comment (rmacklin) [inline app/controllers/images_controller.rb]: Oh, I see you have a global `RecordNotFound` handler in `ApplicationController`. I'd be careful with that, since all controllers are going to inherit that behavior and lots of things could cause a `RecordNotFound`. For that reason I think `rescue_from` is best used in individual controllers (if at all) rather than in `ApplicationController`.

File: app/components/api.js [hunk 15]
```diff
@@ -5,7 +5,7 @@ const HOSTNAME = `${window.location.protocol}//${window.location.host}`;
  */
 class ValidationError extends Error {
   constructor(errors) {
-    super('Validation Error');
+    super();
```
Comment (rmacklin) [inline app/components/api.js]: What was the reason for this change? It's not clear how this relates to "Add API wrapper for image share"

File: app/components/api.js [hunk 16]
```diff
@@ -71,3 +74,11 @@ export function getImage(imageId) {
 export function createImage(data) {
   return post('/images', data).then(d => d.image);
 }
+
+/**
+ * Share an image. Returns a Promise that will be resolved if the share was successful
+ */
+export function shareImage(imageId, recipient, message) {
```
Comment (rmacklin) [inline app/components/api.js]: Did you consider the signature `shareImage(imageId, {recipient, message})` instead? It would make invocations self-documenting, reduce connascence of position to connascence of name, and allow passing an object when you have one (rather than destructuring and then passing as positional arguments)

File: app/components/images/share_dialog.js [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+    const validationErrors = this.state.errors || {};
+
+    let body;
+    if (sent) {
+      body = <p>The image was shared successfully</p>;
+    } else {
+      body = (
+        <form className="form-horizontal" role="form" onSubmit={this.onFormSubmit}>
+
+          {error && <HelpBlock>An unexpected error occurred: {error.message}</HelpBlock>}
```
Comment (rmacklin) [inline app/components/images/share_dialog.js]: screen shot 2016-08-10 at 1 36 57 pm

Do you think "Internal Server Error" is helpful for users?

**Comment by rmacklin** [comment]
Did you test your review app? I can't share an image

**Comment by rmacklin** [comment]
I noticed the share button is only available on the show page. I think it would provide better UX if this action was available on the index page as well. What do you think?

File: test/modal_fix.js [hunk 18]
```diff
@@ -0,0 +1,14 @@
+/* eslint-disable no-underscore-dangle */
```
Comment (rmacklin) [inline test/modal_fix.js]: Can you explain how this file is used?

**Comment by rmacklin** [comment]
This (and other features) should have some full stack integration tests as well. I've opened #30 for you to look into that.

File: config/environment.rb (hunk 1 above)
Comment (tlconnor) [inline config/environment.rb]: I put this here as per the instructions provided by Heroku https://devcenter.heroku.com/articles/sendgrid#ruby-rails

The use of Sendgrid isn't an environment specific configuration.  The only reason it is production only right now is because email sending is disabled in dev / test.  If we did want to send email from dev / test we would then just need to set the Sendgrid environment variables.

File: config/environments/development.rb (hunk 2 above)
Comment (tlconnor) [inline config/environments/development.rb]: This shouldn't have been in development.rb anyway.  I have removed it from development, set it to a mock value in test, and set it via an environment variable in production.

File: config/environments/production.rb (hunk 3 above)
Comment (tlconnor) [inline config/environments/production.rb]: Looks like your comments on development.rb / production.rb were swapped?  I've removed the hostname config for development and set via an environment variable for production.

File: test/mailers/images_mailer_test.rb (hunk 4 above)
Comment (tlconnor) [inline test/mailers/images_mailer_test.rb]: Added

File: app/views/images_mailer/share_image.html.erb (hunk 5 above)
Comment (tlconnor) [inline app/views/images_mailer/share_image.html.erb]: Or even better it should be added to a translation file for i18n, but that would be beyond the scope of this change.

File: app/validators/image_share_validator.rb (hunk 6 above)
Comment (tlconnor) [inline app/validators/image_share_validator.rb]: Yeah looks like you are right.  I didn't think ActiveModel provided a constructor to allow passing attributes to `.new` but it seems that it does.  Now fixed.

File: app/validators/image_share_validator.rb (hunk 7 above)
Comment (tlconnor) [inline app/validators/image_share_validator.rb]: Good point, renamed.

File: test/validators/image_share_validator_test.rb (hunk 8 above)
Comment (tlconnor) [inline test/validators/image_share_validator_test.rb]: added

File: app/controllers/images_controller.rb (hunk 9 above)
Comment (tlconnor) [inline app/controllers/images_controller.rb]: Yeah ok, fixed.

File: app/controllers/images_controller.rb (hunk 10 above)
Comment (tlconnor) [inline app/controllers/images_controller.rb]: Yeah, I know the difference, just getting used to ruby again.  Now fixed.

File: test/controllers/images_controller_test.rb (hunk 11 above)
Comment (tlconnor) [inline test/controllers/images_controller_test.rb]: Both added

File: test/controllers/images_controller_test.rb (hunk 12 above)
Comment (tlconnor) [inline test/controllers/images_controller_test.rb]: As far as I know all the `xhr` helper does is add the `Accept: application/json` HTTP header.  This makes sense for traditional rails apps that respond with HTML by default and require the additional header to specify that the client wants a JSON response.

In this ropes app I'm using Rails 5 in API mode.  This changes the default behavior to respond with JSON rather than HTML, making the `xhr` helper unnecessary.

File: app/controllers/images_controller.rb (hunk 13 above)
Comment (tlconnor) [inline app/controllers/images_controller.rb]: The default behavior in rails when there is an uncaught `RecordNotFound` is to render `404.html`.  In API mode it works a bit differently and instead serializes the exception as JSON.

I didn't want it to do this so I instead added the custom handler to return the JSON I wanted.  If it were not global than 404 errors on other controllers would revert to the default behavior.

A better solution here would be to use `config.exceptions_app` and write an exceptions application to handle any uncaught exceptions we want specific behavior for.  Going that far seemed a bit unnecessary for the ropes app.

File: test/controllers/images_controller_test.rb (hunk 12 above)
Comment (rmacklin) [inline test/controllers/images_controller_test.rb]: Actually, `xhr` sets the `X-Requested-With` header which can be checked in a conditional expression in a controller via `request.xhr?`. So there is a difference. But, I realized you're using `fetch` (rather than XmlHttpRequest) which doesn't set that header, so `xhr` should not be used.

File: test/controllers/images_controller_test.rb (hunk 14 above)
Comment (tlconnor) [inline test/controllers/images_controller_test.rb]: added

File: app/components/api.js (hunk 15 above)
Comment (tlconnor) [inline app/components/api.js]: The string passed in the constructor was not necessary so I removed it.  You are right that it is not related to the commit it was included in, I missed this one when splitting the changes into atomic commits.

File: config/environments/development.rb (hunk 2 above)
Comment (rmacklin) [inline config/environments/development.rb]: You still want it set to something valid (e.g. localhost:5000) in development, otherwise attempts to send any emails with URLs in them (such as the [share image mailer](https://github.com/AppFolioOnboarding/image-sharer-tlconnor/blob/217d965b69f9991926af63770672c172685e8838/app/views/images_mailer/share_image.html.erb#L10)) will blow up with `ActionView::Template::Error: Missing host to link to! Please provide the :host parameter, set default_url_options[:host], or set :only_path to true`

File: app/components/api.js (hunk 16 above)
Comment (tlconnor) [inline app/components/api.js]: If there had been more arguments I would have considered this approach.  With there being only `imageId`, `recipient`, and `message` (with `imageId` and `recipient` being mandatory) I didn't think there was much point.

It is something that could be easily refactored in the future if extra parameters were required.

File: app/components/images/share_dialog.js (hunk 17 above)
Comment (tlconnor) [inline app/components/images/share_dialog.js]: That would depend on the intended audience of the app but for the sake of this exercise I've changed it to "An unexpected error occurred, please try again"

File: test/modal_fix.js (hunk 18 above)
Comment (tlconnor) [inline test/modal_fix.js]: I've added comments to the file.  Its basically a helper to allow the contents of modal dialogs to be referenced from within unit tests.

File: config/environments/development.rb (hunk 2 above)
Comment (tlconnor) [inline config/environments/development.rb]: Good point, added.

File: .rubocop.yml [hunk 19]
```diff
@@ -13,6 +13,9 @@ Metrics/LineLength:
 Metrics/MethodLength:
   Max: 128
 
+Metrics/ClassLength:
+  Max: 1024
```
Comment (bboe) [inline .rubocop.yml]: Is that a reasonable class length? That's a lot of lines for a class if it has a single responsibility.

Comment (tlconnor) [inline .rubocop.yml]: I added it because rubocop was complaining about the length of a test, which I would expect to be able to be long.  I took the new value for the rule directly from `apm_bundle`.

Comment (tlconnor) [inline .rubocop.yml]: I've now fixed this to exclude `test/**/*` rather than increasing the max length.

Comment (bboe) [inline .rubocop.yml]: Can you instead add a `.rubocop.yml` to the test directory that inherits from its parent-directory `.rubocop.yml` and disables any metrics that do not pass?

Comment (rmacklin) [inline .rubocop.yml]: Just as a note re:

> I took the new value for the rule directly from apm_bundle

apm_bundle isn't the best resource for suggested rubocop rules as it is a huge repo with tons of legacy code, so the maximums you'll find there can often be quite large compared to what we'd want to use in a new app

**Comment by rmacklin** [comment]
I noticed the review app has no images. Are your seeds set up properly?

**Comment by rmacklin** [comment]
...And I can't create an image either. Looks like your review apps aren't properly running the migrations and setting up the seeds. Can you fix that and reply here when the review app is working again?

**Comment by tlconnor** [comment]
@rmacklin @bboe the reason the review app didn't work is because I rebased against the latest master which included some additional migrations.  Because this branch included later migrations the older ones were ignored and caused database errors.

I've manually run `db:migrate` which has solved the problem.

**Comment by rmacklin** [comment]
When I share a deleted image I get:

screen shot 2016-08-16 at 5 26 42 pm

The not found error should be handled explicitly. You can redirect to the index page (with a flash message) which prevents the user from repeating the error.

**Comment by rmacklin** [comment]
Can you add the ability to share images from the index page? Having to click into each image to get to the share button is an unnecessary step for users.

**Comment by rmacklin** [comment]
Refactors like 78224b27e166fb8ee2cfe12fc39ad03630d17b7e should be done on their own branches to keep the scope of each PR focused (i.e. apply the single responsibility principle to branches, just like you would to commits) and to deliver value as fast as possible.

File: app/components/api.js (hunk 15 above)
Comment (rmacklin) [inline app/components/api.js]: Since you'll be rebasing anyway (https://github.com/AppFolioOnboarding/image-sharer-tlconnor/pull/29#issuecomment-240282089), it would be nice to split this out into its own commit.

**Comment by tlconnor** [comment]
@rmacklin @bboe 

I have created a new PR https://github.com/AppFolioOnboarding/image-sharer-tlconnor/pull/33 that adds functionality somewhat similar to the rails "flash".  Once that is merged I can use it to solve the issue @rmacklin noted when sharing deleted images.

Regarding sharing images from the index page, this is design decision that is a matter of opinion. I think it is better to not be on the index page as it would clutter the page with too many buttons. Regardless of our opinions it is outside the scope of a code review unless explicitly defined in the original requirements.

**Comment by rmacklin** [comment]
That's normally true if I was only acting as a developer in my reviews, but I'm also acting as the UX and QA members which would normally be looking at your branches as well. In this case the UX argument for having it on the index page is a strong one: by adding the unnecessary step, you're going to cause fewer users to "convert". (Obviously this is all contrived, but it simulates concerns working on a real application. This type of back and forth during feature development is normal.)

Additionally, if you're really concerned about too many buttons, you could avoid it by exposing the per-image actions on hover.

**Comment by tlconnor** [comment]
@rmacklin @bboe 

I would also like to share my view on the single responsibility principle. In this PR the principle is being treated as a concrete rule that is being taken to the extreme.  I would hope this is just due to ropes being a learning exercise and a way to teach new engineers how to organize commits / PRs.

I agree that a clean commit history is important, and to a certain extent having a single responsibility within a PR is important for easier reviews.  These are never concrete rules though and a balance needs to be made between the gain from applying these principles and the additional effort involved.

Take 78224b2 for example.  The commit refactors the "delete image" modal into a new react component.  The commit does appear to have nothing to do with the "share image" use case and you could argue that it should be a separate PR.  What really happened here is the addition of the "share image" dialog caused the delete dialog unit tests to fail.  I decided to perform the refactor to add code consistency and to make the tests more reliable (and non conflicting).  To achieve this I had three choices:

1. fix the failing test in a temporary way, then perform the refactor in a separate branch
2. switch to a new branch, perform the refactor (including updating the test), submit a PR, wait for the PR to be merged, then rebase the share image PR
3. perform the refactor in a new commit but as part of the same PR

Of these three options we have the following pros and cons:

1. PROS: cleaner PRs (apart from the test change), CONS: not so clean commit history (test is ultimately updated twice)
2. PROS: cleaner PRs, clean commit history, CONS: more developer effort and slower turnaround
3. PROS: clean commit history, quicker turnaround, less effort, CONS: violates single responsibility principle for the PR

Given that a PR is a temporary entity I see option #3 as being the best.  Overall effort (and therefore cost) across both dev and reviewer is the least and the final commit history is still clean.

If this were really a learning exercise I could see the reason to split out 78224b2, 3920491, and 61f81f0 as separate PRs however in a real world use case the extra effort would be counter productive.

**Comment by rmacklin** [comment]
> to a certain extent having a single responsibility within a PR is important for easier reviews.

This greatly over-trivializes the importance. It doesn't just make the review easier, it makes the review more likely to catch errors that would have been missed if everything was grouped into one PR. (The same is true in the development phase). More importantly, in our real applications there are many hands in the pot. Teams do not have split repositories or code ownership. Many teams are making changes in the same area. By pushing a refactor into a feature branch, you force those two independent changes to be merged together and you increase the turnaround time for that one branch, which makes it all the more likely that before this gets merged, someone else has merged changes that conflict with your refactor, and now you've suddenly got to deal with resolving those (which are often nontrivial) before you can deliver the sharing feature to customers, even if there is nothing holding up the sharing feature. And that's not all. Suppose the share feature has a serious bug and we have to patch production. That should be as easy as first reverting the PR (and quickly unbreaking the application), and then spending time fixing the problem and opening a new PR. But suddenly you're now reverting the refactor as well, which again could have serious merge conflicts because other developers have made changes in that area, and you've got to deal with that even if there were no conflicts in the sharing changes. I'm not making this stuff up; it's happened here numerous times. This is _precisely_ why we have Ropes. (You're absolutely correct that it's a learning exercise.) It grew out of a need for this kind of guided learning because of real issues we experienced repeatedly.

I'll respond to the rest of your comment in a subsequent reply.

**Comment by bboe** [comment]
I just want to say that this is a great discussion. @tlconnor you bring up some really good points, and it's clear that you've thought through some various approaches. I agree with @rmacklin's assessment that making separate pull requests (approach 2) is the way to go for the reasons he described. While you are correct that it does provide extra overhead, that small percentage of a time where something goes wrong is worth the extra effort.

**Comment by rmacklin** [comment]
I disagree with some of your pros/cons (also without the full context I can't be sure those are the only three options, but I'll assume they are).

> 1. CONS: not so clean commit history (test is ultimately updated twice)

It's updated twice, but on different branches. I don't think that makes the commit history unclean. An unclean history would be when something is done, then undone or significantly reworked just a few commits later in the same PR. Additionally, the first update is expected to be minimal.

> 2. CONS: more developer effort and slower turnaround

The extra effort here is minimal (as @bboe mentioned). It's a matter of creating a new branch (and maybe reordering a commit in the history prior to that), and opening a new PR for it. Rebasing after the first branch is merged isn't even technically necessary. If you just want to get those refactor commits to not show on the share PR you can simply close it and open a new PR and they'll be gone since they're in the target branch.

Additionally, the "slower turnaround" point is flawed. For one, the combined turnaround shouldn't be significantly longer. And moreover, value is delivered _sooner_ here (and also in the first approach) because you have two merges, rather than one big shebang.

> 3. PROS: clean commit history, quicker turnaround, less effort

As discussed, I disagree with these pros.

> 3. CONS: violates single responsibility principle for the PR

As mentioned in my previous reply, there are a lot more cons.

> Overall effort (and therefore cost) across both dev and reviewer is the least

I disagree that overall effort of the reviewer is the least in option 3. In a real world scenario, it's likely that you'd actually be best suited getting two separate reviewers for the refactor and the feature, because the person reviewing the feature isn't the most qualified to review the refactor given that that's someone else's area of expertise. When it's all in one branch, you're unlikely to seek out the second reviewer for those changes, and consequently you'll be making the feature reviewer spend time understanding the code being refactored - which isn't necessarily a bad thing but it will certainly take that reviewer more time and effort (or, if the reviewer doesn't have that time to spend, they'll ask you to split those changes into a separate PR and get the more qualified person to review it - not unlike what I did here).

I see option 1 (or at least a modified version of it) as the winner. It's better than option 2 because option 2 forces the share branch to wait on the refactor getting merged. If you make a quick fix for the test, you can proceed implementing the share feature to completion, and while you're waiting for that to be reviewed/QAed/etc, you can work on the refactor to completion, and these branches will be independent. That said, I don't think option 2 is _much_ worse.

(I don't think I need to reiterate why I think option 3 is the worst option.)

**Comment by rmacklin** [comment]
> I just want to say that this is a great discussion.

Completely agree! I want to say thanks for sparking this discussion. It got me to vocalize some thoughts that have been in my head from recently experiencing very similar situations in our real apps. So thanks!

**Comment by tlconnor** [comment]
I'd say what this comes down to is the difference between working in this theoretical assignment vs a real world application.

I agree with all of the points @rmacklin makes with regard to working in large teams and they are all very real issues that I have experienced many times in the past (on engineering teams both larger and smaller than we have here at AppFolio).

Above all I am a pragmatist and always consider the implications of any decision.  For this reason I have a strong aversion to blanket rules and prefer to judge decisions based on the real world consequences and merits.

In this specific example I don't see this as a learning exercise, there is nobody else working on the codebase, and no chance of conflicts.  I believe my choices were appropriate given all of this.

Thanks for all of your feedback so far, it is indeed a good discussion and it is certainly not my intention to cause conflict or unintentional argument.

**Comment by tlconnor** [comment]
I've fixed the "share deleted image" issue

**Comment by tlconnor** [comment]
@rmacklin @bboe as far as I can tell all functional issues are now resolved.

If the only outstanding issues are now the ones we have been debating I will for the sake of the exercise re-create the changeset as multiple PRs.

**Comment by tlconnor** [comment]
@rmacklin @bboe I have rebased this PR against the latest master so the refactors are now gone.

File: app.json [hunk 20]
```diff
@@ -18,6 +21,12 @@
...
       "generator": "secret"
+    },
+    "SENDGRID_USERNAME": {
+      "required": true
+    },
+    "SENDGRID_PASSWORD": {
+      "required": true
     }
   },
   "addons": [
```
Comment (rmacklin) [inline app.json]: The SendGrid addon should be declared here too.

Comment (tlconnor) [inline app.json]: added

**Comment by bboe** [comment]
Closing this to remove it from my list of mentioned pull requests.
